scripts/validate-oyd-vnet.sh

```diff
@@ -0,0 +1,368 @@
+
+# verify resource types match expectation
+if [[ $(get_resource_type_from_resource "$acs_resource") != "Microsoft.Search/searchServices" ]]; then
+    echo "The value given for --azure-search-resource-id is not a valid Azure Search resource ID."
+    exit 1
+elif [[ $(get_resource_type_from_resource "$aoai_resource") != "Microsoft.CognitiveServices/accounts" ]]; then
+    echo "The value given for --azure-openai-resource-id is not a valid Azure OpenAI resource ID."
+    exit 1
+elif [[ $(get_resource_type_from_resource "$storage_resource") != "Microsoft.Storage/storageAccounts" ]]; then
+    echo "The value given for --storage-account-resource-id is not a valid storage account resource ID."
+    exit 1
+elif [[ $(get_resource_type_from_resource "$virtual_network_resource") != "Microsoft.Network/virtualNetworks" ]]; then
+    echo $virtual_network_resource | jq
+    echo "The value given for --virtual-network-resource-id is not a valid Azure Search resource ID."
+    exit 1
+elif [[ $(get_resource_type_from_resource "$gateway_resource") != "Microsoft.Network/virtualNetworkGateways" ]]; then
+    echo "The value given for --gateway-resource-id is not a valid Azure Search resource ID."
+    exit 1
+fi
+
+# verify trusted services configuration.
+check_bypass_azure_services_configuration "$storage_resource"
+check_bypass_azure_services_configuration "$aoai_resource"
+
+# verify network topology (vnet + private endpoints configuration).
+# 1) VNET has private endpoint connection to each resource, and each are approved by the user.
+check_resources_subnet_connection $subscription_id "$virtual_network_resource" "$data_resources"
+
+# 2) VNET also has a VPN gateway attached to it.
+check_gateway_subnet_connection $virtual_network_resource_id "$gateway_resource"
+
+# 3) Additional private endpoint connection to AME exists on search resource.
+check_ame_private_endpoint_connection "$acs_resource"
+
+# verify roles and scopes on each item.
+check_mi_roles $subscription_id "$acs_resource" "[\"Cognitive Services OpenAI Contributor\",\"Storage Blob Data Contributor\"]"
+check_role_assignment_scope $subscription_id "$acs_resource" "Cognitive Services OpenAI Contributor" $aoai_resource_id
+check_role_assignment_scope $subscription_id "$acs_resource" "Storage Blob Data Contributor" $storage_account_resource_id
+check_mi_roles $subscription_id "$aoai_resource" "[\"Search Index Data Reader\",\"Search Service Contributor\",\"Storage Blob Data Contributor\"]"
+check_role_assignment_scope  $subscription_id "$aoai_resource" "Search Index Data Reader" $search_resource_id
+check_role_assignment_scope  $subscription_id "$aoai_resource" "Search Service Contributor" $search_resource_id
+check_role_assignment_scope  $subscription_id "$aoai_resource" "Storage Blob Data Contributor" $storage_account_resource_id
+
+# check that ACS resource has both API key + RBAC enabled
+check_rbac "$acs_resource"
```

Comment:
can we also check the login user to resources role assignments? Above only check service to service permissions. We also need to check user to service permissions.

Reply:
Yep, I have added this in the last iteration.  Let me know if there is anything missed.